src/main/java/io/cdap/plugin/sap/SapODataConfig.java

```diff
@@ -0,0 +1,251 @@
+
+  @Name(SapODataConstants.QUERY)
+  @Description("OData query options to filter the data.")
+  @Macro
+  @Nullable
+  private String query;
+
+  @Name(SapODataConstants.USERNAME)
+  @Description("Username for basic authentication.")
+  @Macro
+  @Nullable
+  private String user;
+
+  @Name(SapODataConstants.PASSWORD)
+  @Description("Password for basic authentication.")
+  @Macro
+  @Nullable
+  private String password;
+
+  @Name(SapODataConstants.SCHEMA)
+  @Description("Schema of records output by the source.")
+  @Nullable
+  private String schema;
+
+  public SapODataConfig(String referenceName, String url, String resourcePath, String query, String user,
+                        String password, String schema) {
+    this.referenceName = referenceName;
+    this.url = url;
+    this.resourcePath = resourcePath;
+    this.query = query;
+    this.user = user;
+    this.password = password;
+    this.schema = schema;
+  }
+
+  public String getReferenceName() {
+    return referenceName;
+  }
+
+  public String getUrl() {
+    return url;
+  }
+
+  public String getResourcePath() {
+    return resourcePath;
+  }
+
+  @Nullable
+  public String getQuery() {
+    return query;
+  }
+
+  /**
+   * An OData query can contain '$select' option. The $select option specifies a subset of properties to include in the
+   * response body. For example, to get only the name and price of each product, the following query can be used:
+   * 'http://localhost/odata/Products?$select=Price,Name'.
+   *
+   * @return empty list if no '$select' query option specified. List of the property names to include, otherwise(order
+   * is preserved).
+   */
+  public List<String> getSelectProperties() {
+    if (Strings.isNullOrEmpty(query) || !query.contains("$select")) {
+      return Collections.emptyList();
+    }
+    // There is no straightforward way to parse $select query options using Olingo V2
+    String selectOption = "$select=";
+    int start = query.indexOf(selectOption) + selectOption.length();
+    int end = query.indexOf("&", start);
+    String commaSeparatedPropertyNames = end != -1 ? query.substring(start, end) : query.substring(start);
+
+    return Arrays.asList(commaSeparatedPropertyNames.split(","));
+  }
+
+  @Nullable
+  public String getUser() {
+    return user;
+  }
+
+  @Nullable
+  public String getPassword() {
+    return password;
+  }
+
+  @Nullable
+  public String getSchema() {
+    return schema;
+  }
+
+  @Nullable
+  public Schema getParsedSchema() {
+    try {
+      return schema == null ? null : Schema.parseJson(schema);
+    } catch (IOException e) {
+      throw new InvalidConfigPropertyException("Invalid schema", e, SapODataConstants.SCHEMA);
+    }
+  }
+
+  /**
+   * Validates {@link SapODataConfig} instance.
+   */
+  public void validate() {
+    if (Strings.isNullOrEmpty(referenceName)) {
+      throw new InvalidConfigPropertyException("Reference name must be specified", Constants.Reference.REFERENCE_NAME);
```

Comment:
CDAP 6.1.0-SNAPSHOT now has a way to capture all of these validation errors at once. This project should be updated to use the newest version and use the new FailureCollector instead of throwing InvalidConfigPropertyException.

Reply:
Alternatively, this can be done in a follow up PR.

Reply:
Changed to use `FailureCollector`. Tests updated.